### PR TITLE
CI: enable testing with coming PT 2.2

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

- [x] building images for new PT
- [x] upload the new images
- [x] use these images in CI
- [ ] if all pass ad it to GC

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--19289.org.readthedocs.build/en/19289/

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

**Comment by Borda** [comment]
@awaelchli what is your suggestion about
```
E       ValueError: Does not validate against any of the Union subtypes
E       Subtypes: (<class 'torch.optim.lr_scheduler.StepLR'>, <class 'torch.optim.lr_scheduler.ExponentialLR'>)
E       Errors:
E         - Problem with given class_path 'torch.optim.lr_scheduler.StepLR':
E             Parser key "verbose":
E               Expected a <class 'bool'>. Got value: deprecated
E         - Import path torch.optim.lr_scheduler.StepLR does not correspond to a subclass of <class 'torch.optim.lr_scheduler.ExponentialLR'>
E       Given value type: <class 'jsonargparse._namespace.Namespace'>
E       Given value: Namespace(class_path='torch.optim.lr_scheduler.StepLR', init_args=Namespace(step_size=50))

**Comment by awaelchli** [comment]
@carmocca @mauvilsa PyTorch changed the [default value of the "verbose" argument of LRSchedulers](https://github.com/pytorch/pytorch/blob/13473df0d770e67f92cb4bb1b728f3746f160b7c/torch/optim/lr_scheduler.py#L35) to the string "deprecated" but didn't update the type (bool). Now jsonargparse can't validate the schedulers. Do you know a way around this or is the only option to fix it in PyTorch?

**Comment by mauvilsa** [comment]
> One change that could be done in jsonargparse that kind of makes sense is: when validating if the given value is the same as the default, then don't mind the invalid type. Would be better if people don't set invalid defaults. But since it is the default, then not bad considering it as valid.

jsonargparse v4.27.2 is now available including this change. You can try it out here.

**Comment by Borda** [comment]
Not sure what is happening but seeing with Fabric some `torch.distributed.DistNetworkError: Unknown error`

**Comment by awaelchli** [comment]
@mauvilsa Thanks for making the change so quickly. We've pulled in the latest release here but the tests still fail. I looked into it for a while and tried changing the definitions [here](https://github.com/Lightning-AI/pytorch-lightning/blob/d02009af76531e803a6e1593827894b436e366fc/src/lightning/pytorch/utilities/types.py#L80-L83) because it looks like it's not able to infer that the default comes from the super class, but to no avail. Is there a way to give jsonargparse "fake" classes to validate against, where we could provide the updated scheduler type annotation?

I'm considering sending a PR to PyTorch, but at the same time we need to find a workaround for the current version 2.2.

**Comment by mauvilsa** [comment]
@awaelchli I underestimated the required changes. Since pytorch 2.2 isn't out yet, I didn't test with the actual code. I only updated the logic related to parsing, but now I see it is also required to update the instantiation and the serialization. I will look into it.

**Comment by mauvilsa** [comment]
I implemented a fix in https://github.com/omni-us/jsonargparse/pull/435. I tested manually by modifying my pytorch installation. If you like, before creating a jsonargparse release, it could be tested here by temporarily changing the requirement to `jsonargparse[signatures] @ https://github.com/omni-us/jsonargparse/zipball/main`.

**Comment by mauvilsa** [comment]
It still doesn't work, doesn't it? How can I install a pre-release of pytorch 2.2?

**Comment by awaelchli** [comment]
Hey @mauvilsa Yes it seems so. Here are the instructions for different platforms to install the 2.2 rc:

> PyTorch Release 2.2.0 - Final RC is available
> PyTorch core and Domain Libraries are available for download from pytorch-test channel.
>
> Reminder of key dates:
> * M5: External-Facing Content Finalized (1/26/24)
> * M6: Release Day (1/30/24)
>
> Following are instructions on how to download different versions of RC for testing.
> CONDA CPU:
> Windows/LInux:
> conda install pytorch torchvision torchaudio cpuonly -c pytorch-test
> MacOS:
> conda install pytorch torchvision torchaudio -c pytorch-test
> CUDA 11.8, 12.1
> Windows/Linux:
> conda install pytorch torchvision torchaudio pytorch-cuda=11.8 -c pytorch-test -c nvidia
> conda install pytorch torchvision torchaudio pytorch-cuda=12.1 -c pytorch-test -c nvidia
>
> PIP CPU:
> Windows/Linux/MacOS:
> pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/test/cpu
>
> PIP CUDA 11.8, 12.1
> Windows/Linux:
> pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/test/cu118
> pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/test/cu121
>
> PIP ROCM 5.7:
> pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/test/rocm5.7/
>
> PIP Linux Aarch64:
> pip3 install torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/test/cpu

**Comment by mauvilsa** [comment]
I am kind of lost. The tests run fine for me locally. What I did was, first checkout this branch and then:
```
pip3 install -r requirements/pytorch/base.txt
pip3 install -r requirements/pytorch/test.txt
export PACKAGE_NAME=pytorch; pip3 install -e .
pip3 install torch-2.2.0.dev20231206+cpu.cxx11.abi-cp310-cp310-linux_x86_64.whl
pytest -v tests/tests_pytorch/test_cli.py
```

All tests passed. Only one skipped because torchvision is not installed. But the tests related to schedulers are run.

**Comment by awaelchli** [comment]
@mauvilsa You are right, I did the same locally installing jsonargparse from main branch and the tests passed. I must have messed up something installing jsonargparse from main here in the PR.

File: requirements/pytorch/extra.txt [hunk 1]
```diff
@@ -5,7 +5,8 @@
 matplotlib>3.1, <3.9.0
 omegaconf >=2.0.5, <2.4.0
 hydra-core >=1.0.5, <1.4.0
-jsonargparse[signatures] >=4.26.1, <4.27.0
+jsonargparse[signatures] @ git+https://github.com/omni-us/jsonargparse
```
Comment (Borda) [inline requirements/pytorch/extra.txt:8]:
```suggestion
jsonargparse[signatures] @ https://github.com/omni-us/jsonargparse/archive/main.zip
```

**Comment by awaelchli** [comment]
I'm also lost. What's the difference local vs here in the CI 🤔 Tested in clean env with Python 3.10, PyTorch 2.2 and jsonargparse from main, no issues.

**Comment by awaelchli** [comment]
Ok this seems to be just a stupid installation problem. I did what @carmocca said and logged in via SSH into the job. Running the tests there failed, but when I manually cloned jsonargparse and ran `pip install -e .[signatures]` then the tests started passing. So I think as soon as @mauvilsa will release the next version, we should be fine.

File: .github/workflows/ci-tests-pytorch.yml [hunk 2]
```diff
@@ -49,13 +49,17 @@ jobs:
...
           - { os: "macOS-11", pkg-name: "lightning", python-version: "3.10", pytorch-version: "2.0" }
           - { os: "ubuntu-20.04", pkg-name: "lightning", python-version: "3.10", pytorch-version: "2.0" }
           - { os: "windows-2022", pkg-name: "lightning", python-version: "3.10", pytorch-version: "2.0" }
           - { os: "macOS-11", pkg-name: "lightning", python-version: "3.10", pytorch-version: "2.1" }
           - { os: "ubuntu-20.04", pkg-name: "lightning", python-version: "3.10", pytorch-version: "2.1" }
           - { os: "windows-2022", pkg-name: "lightning", python-version: "3.10", pytorch-version: "2.1" }
+          # only run PyTorch future
+          - { os: "macOS-12", pkg-name: "lightning", python-version: "3.10", pytorch-version: "2.2" }
+          - { os: "ubuntu-22.04", pkg-name: "lightning", python-version: "3.10", pytorch-version: "2.2" }
+          - { os: "windows-2022", pkg-name: "lightning", python-version: "3.10", pytorch-version: "2.2" }
```
Comment (awaelchli) [inline .github/workflows/ci-tests-pytorch.yml:62]: maybe we should set Python 3.11 here?

**Review by awaelchli — APPROVED** [review]

**Comment by awaelchli** [comment]
Thank you very much @mauvilsa! Everything is working now 🎉

File: tests/tests_fabric/plugins/precision/test_amp_integration.py [hunk 3]
```diff
@@ -37,6 +40,11 @@ def forward(self, x):
         return output
 
 
+@pytest.mark.xfail(
+    # https://github.com/pytorch/pytorch/issues/116056
+    sys.platform == "win32" and _TORCH_GREATER_EQUAL_2_2,
```
Comment (Borda) [inline tests/tests_fabric/plugins/precision/test_amp_integration.py:45]: wow, did not know that `xfail` has also conditions :rocket:

**Review by carmocca — APPROVED** [review]

File: .github/workflows/ci-tests-pytorch.yml (hunk 2 above)
Comment (Borda) [reply .github/workflows/ci-tests-pytorch.yml:62]: I tries and still some packages does not have it compiled

**Pull request merged**